### PR TITLE
Accept nondifferentiated args

### PR DESCRIPTION
**Pull request by thorek1**
The specific use case is that RuntimeGeneratedFunctions does not accept keyword arguments and in that case the construction of an implicit function becomes unnecessarily bloated.

Overall it's a generalisation of the interface.

I added tests, docs, and checked it with my package and it simplifies things a lot.

### CODE REVIEW

File: docs/src/faq.md [hunk 1]
```diff
@@ -33,8 +33,8 @@ The forward mapping _and_ conditions need arrays: for example, instead of return
 
 ## Multiple inputs / outputs
 
-In this package, implicit functions can only take a single input array `x` and output a single output array `y` (plus the byproduct `z`).
-But sometimes, your forward mapping or conditions may require multiple input arrays, say `a` and `b`:
+Implicit functions have as their first positional argument an input array `x` which contains the elements to be differentiated. Implicit functions also accept further positional arguments and keyword arguments. The same goes for the conditions, as in their first two (three) positional arguments must be `x` and the output of the forward mapping  `y` (plus an optional byproduct `z`). The conditions must accept the same further positional (following the previously mentioned arguments) and keyword arguments as the forward mapping.
```
Comment (gdalle) [inline docs/src/faq.md:36]: This goes against the ChainRules convention of differentiating with respect to all arguments and [not kwargs](https://juliadiff.org/ChainRulesCore.jl/stable/FAQ.html#Where-are-the-derivatives-for-keyword-arguments?)

File: ext/ImplicitDifferentiationChainRulesExt.jl [hunk 2]
```diff
@@ -17,11 +17,11 @@ We compute the vector-Jacobian product `Jᵀv` by solving `Aᵀu = v` and settin
...
-    rc::RuleConfig, implicit::ImplicitFunction, x::AbstractArray{R}; kwargs...
+    rc::RuleConfig, implicit::ImplicitFunction, x::AbstractArray{R}, args...; kwargs...
 ) where {R}
-    y_or_yz = implicit(x; kwargs...)
+    y_or_yz = implicit(x, args...; kwargs...)
     backend = reverse_conditions_backend(rc, implicit)
-    Aᵀ_op, Bᵀ_op = reverse_operators(backend, implicit, x, y_or_yz; kwargs)
+    Aᵀ_op, Bᵀ_op = reverse_operators(backend, implicit, x, y_or_yz, args; kwargs)
     byproduct = y_or_yz isa Tuple
     implicit_pullback = ImplicitPullback{byproduct}(Aᵀ_op, Bᵀ_op, implicit.linear_solver, x)
```
Comment (gdalle) [inline ext/ImplicitDifferentiationChainRulesExt.jl:26]: This becomes wrong cause we need to return a tangent for each `arg`

File: test/systematic.jl [hunk 3]
```diff
@@ -6,22 +6,22 @@ using ForwardDiff: ForwardDiff
 import ImplicitDifferentiation as ID
 using ImplicitDifferentiation: ImplicitFunction, identity_break_autodiff
 using ImplicitDifferentiation: DirectLinearSolver, IterativeLinearSolver
-using JET
+# using JET
```
Comment (gdalle) [inline test/systematic.jl:9]: Oh I see, so now we're just deactivating tests huh ;)

File: test/systematic.jl [hunk 4]
```diff
@@ -6,22 +6,22 @@ using ForwardDiff: ForwardDiff
...
 using LinearAlgebra
 using Random
 using ReverseDiff: ReverseDiff
 using StaticArrays
 using Test
 using Zygote: Zygote, ZygoteRuleConfig
 
-@static if VERSION < v"1.9"
+# @static if VERSION < v"1.9"
```
Comment (gdalle) [inline test/systematic.jl:17]: I SEE YOU

File: test/systematic.jl [hunk 5]
```diff
@@ -38,50 +38,61 @@ function mysqrt(x::AbstractArray)
     return sqrt.(identity_break_autodiff(x))
 end
 
-function make_implicit_sqrt(; kwargs...)
+function make_implicit_sqrt(args... ; kwargs...)
```
Comment (gdalle) [inline test/systematic.jl:41]: No need for `args...` here

Comment (gdalle) [reply test/systematic.jl:41]: or in any of the cases that follow

File: test/systematic.jl [hunk 6]
```diff
@@ -136,32 +152,39 @@ function test_implicit_rrule(rc, x; kwargs...)
     y1, pb1 = @inferred rrule(rc, imf1, x)
     (y2, z2), pb2 = @inferred rrule(rc, imf2, x)
     y3, pb3 = @inferred rrule(rc, imf3, x; p=0.5)
+    y4, pb4 = @inferred rrule(rc, imf4, x, 0.5)
 
     dimp1, dx1 = @inferred pb1(dy)
     dimp2, dx2 = @inferred pb2((dy, dz))
     dimp3, dx3 = @inferred pb3(dy)
+    dimp4, dx4 = @inferred pb4(dy)
```
Comment (gdalle) [inline test/systematic.jl:160]: This should be `dimp4, dx4, dargs4...`

**Review by gdalle — COMMENTED** [review]
I don't think this is a good idea because it either
- breaks with the ChainRules.jl convention of differentiating wrt all args
- requires us to write an unknown number of pullbacks

**Comment by gdalle** [comment]
@mohamed82008 wanna weigh in?

**Comment by gdalle** [comment]
I'm also curious what @oxinabox thinks (sorry for the ping): is it morally wrong from a ChainRules perspective to accept positional arguments for which we flat out refuse to compute derivatives?

**Comment by gdalle** [comment]
Once #91 is merged the ChainRules tests will be more robust to work on this PR

**Comment by thorek1** [comment]
I don't see the chain rules issue but that's just my ignorance. Anyway, a solution might be to write an internal wrapper function!?

**Comment by gdalle** [comment]
gonna work on it tomorrow

**Comment by mohamed82008** [comment]
An example showing exactly where the problem is can help us figure out a better solution than this. For instance, you can easily close over variables or define a wrapper function which has a kwarg and passes it on as an arg. There are multiple ways to avoid differentiating wrt an input.

**Comment by thorek1** [comment]
I generate functions using RuntimeGeneratedFunctions.jl which does not allow me to use known arguments or opaque clojures (wrapper functions). I can post an example of how I manage to use ImplicitDifferentiation in that context later.

**Comment by gdalle** [comment]
I've been taking a look @mohamed82008 and it doesn't seem very complicated to make it work, the challenge is to document it well enough. It's a strict feature extension, so it doesn't really hurt the rest. I'll post a working version tomorrow if I can

**Comment by thorek1** [comment]
the ingredient are: a forward mapping which depends on a function generated by RuntimeGeneratedFunctions.jl and that same function is the condition. These are only known at runtime and therefore I have a generic function describing the `ImplicitFunction`

```Julia
block_solver_AD(parameters_and_solved_vars::Vector{<: Real}, 
    n_block::Int, 
    ss_solve_blocks::Function, 
    # ss_solve_blocks_no_transform::Function, 
    # f::OptimizationFunction, 
    guess::Vector{Float64}, 
    lbs::Vector{Float64}, 
    ubs::Vector{Float64};
    tol::AbstractFloat = eps(Float64),
    # timeout = 120,
    starting_points::Vector{Float64} = [0.897, 1.2, .9, .75, 1.5, -.5, 2.0, .25],
    # fail_fast_solvers_only = true,
    verbose::Bool = false) = ImplicitFunction(x -> block_solver(x,
                                                            n_block, 
                                                            ss_solve_blocks,
                                                            # f,
                                                            guess,
                                                            lbs,
                                                            ubs;
                                                            tol = tol,
                                                            # timeout = timeout,
                                                            starting_points = starting_points,
                                                            # fail_fast_solvers_only = fail_fast_solvers_only,
                                                            verbose = verbose)[1],  
                                        (x,y) -> ss_solve_blocks(x,y))
```

and the forward mapping function (`block_solver`) which depends on the ss_solve_blocks function is:
```Julia
function block_solver(parameters_and_solved_vars::Vector{Float64}, 
                        n_block::Int, 
                        ss_solve_blocks::Function, 
                        # SS_optimizer, 
                        # f::OptimizationFunction, 
                        guess::Vector{Float64}, 
                        lbs::Vector{Float64}, 
                        ubs::Vector{Float64};
                        tol::AbstractFloat = eps(),
                        # timeout = 120,
                        starting_points::Vector{Float64} = [0.897, 1.2, .9, .75, 1.5, -.5, 2, .25],
                        # fail_fast_solvers_only = true,
                        verbose::Bool = false)

    ...

    return sol_values, sol_minimum
end
```

as for the generated functions at runtime they look like this:

here is an example of `ss_solve_blocks`:
```Julia
RuntimeGeneratedFunctions.RuntimeGeneratedFunction{(:parameters_and_solved_vars, :guess), MacroModelling.var"#_RGF_ModTag", MacroModelling.var"#_RGF_ModTag", (0x8311c1df, 0xebe928d0, 0x62010178, 0xa0812539, 0xf638389f), Expr}(quote
    #= /Users/thorekockerols/GitHub/MacroModelling.jl/src/MacroModelling.jl:834 =#
    #= /Users/thorekockerols/GitHub/MacroModelling.jl/src/MacroModelling.jl:838 =#
    P = guess[1]
    c = guess[2]
    k = guess[3]
    l = guess[4]
    n = guess[5]
    #= /Users/thorekockerols/GitHub/MacroModelling.jl/src/MacroModelling.jl:839 =#
    alp = parameters_and_solved_vars[1]
    bet = parameters_and_solved_vars[2]
    psi = parameters_and_solved_vars[3]
    del = parameters_and_solved_vars[4]
    #= /Users/thorekockerols/GitHub/MacroModelling.jl/src/MacroModelling.jl:840 =#
    m = parameters_and_solved_vars[5]
    ➕₂ = parameters_and_solved_vars[6]
    ➕₃ = parameters_and_solved_vars[7]
    ➕₄ = parameters_and_solved_vars[8]
    ➕₅ = parameters_and_solved_vars[9]
    #= /Users/thorekockerols/GitHub/MacroModelling.jl/src/MacroModelling.jl:845 =#
    return [(bet * (alp * k ^ (alp - 1) * n ^ (1 - alp) * exp(➕₂) + (1 - del) * exp(➕₃))) / c - 1 / c, P * c - m, (-P * c * psi) / ((1 - n) * (1 - psi)) + l / n, ((c - k * (1 - del) * exp(➕₅)) + k) - k ^ alp * n ^ (1 - alp) * exp(➕₄), (-bet * k ^ alp * n ^ (1 - alp) * (1 - alp) * exp(➕₄)) / (l * m) + 1 / P]
end)
```

and then all of this is made to work in a function generated at runtime the following way:
```Julia
block_solver_RD = block_solver_AD([alp, bet, psi, del, m, ➕₂, ➕₃, ➕₄, ➕₅], 1, 𝓂.ss_solve_blocks[1], inits, lbs, ubs, verbose = verbose)
solution = block_solver_RD([alp, bet, psi, del, m, ➕₂, ➕₃, ➕₄, ➕₅])
```

all of this works with ImplicitDifferentiation v0.3.0

given my constraints I have to rely on these rather ugly constructs and define a function to write and ImplicitFunction which then I call to get the result. It does work but it was non-trivial to figure out and is definitely not elegant. With the code in this branch I can significantly reduce the amount of code and it works for the `ForwardDiff` case.

**Comment by mohamed82008** [comment]
Can you please make the example minimal and runnable but failing due to kwargs?

**Comment by thorek1** [comment]
I'll give it a go tomorrow

**Comment by oxinabox** [comment]
> I'm also curious what @oxinabox thinks (sorry for the ping): is it morally wrong from a ChainRules perspective to accept positional arguments for which we flat out refuse to compute derivatives?

Seems fine to me.
You should return a `NotImplemented` for them from the pullback, via the `@not_implemented` macro in ChainRulesCore.
That explains what is going on. That should ensure that if the user does attempt to AD wrt them they get an informative error message.

**Comment by gdalle** [comment]
>  You should return a NotImplemented for them from the pullback, via the @not_implemented macro in ChainRulesCore.

Not sure I wanna do that cause I find ChainRulesTestUtils.jl very convenient and this breaks it

**Comment by oxinabox** [comment]
> Not sure I wanna do that cause I find ChainRulesTestUtils.jl very convenient and this breaks it ^^

Hmm, can you open an issue about that?
I would have expected that to work.
I suspect the workaround is to specify it as not to be differentiated via passing that argument to the test function with 
`x ⊢NoTangent()`

**Comment by gdalle** [comment]
@thorek1 wanna take this branch for a spin, see if you like it?

**Comment by gdalle** [comment]
@oxinabox I was just referring to this part, which is actually good and expected behavior for CRTU (I think)

> Note: [ChainRulesTestUtils.jl](https://github.com/JuliaDiff/ChainRulesTestUtils.jl) marks @not_implemented tangents as "test broken".

Source: https://juliadiff.org/ChainRulesCore.jl/stable/rule_author/writing_good_rules.html#Use-@not_implemented-appropriately

**Comment by thorek1** [comment]
works like a charm. well done and thanks a lot. this makes coding Implicit diff problems way easier.

**Comment by gdalle** [comment]
If this can be done in a non-breaking way, we'll do it as part of the v0.5 series.

I have some more cleanup to do but I should be able to tag the release this week

**Comment by thorek1** [comment]
yeah should be non-breaking

**Pull request merged**